### ISSUE TITLE
Provide ability to call gaphor console externally

### ISSUE THREAD

**Issue by hulsed** [gaphor/gaphor#2821]
### Is your feature request related to a problem?

I am a researcher who would like to build some tools simulation tools connected to gaphor. One thing I would like to be able to do is call Gaphor's console from an external console (python instance, powershell, jupyter).

Currently, this is impossible without a development installation, which may be difficult or impossible to setup without administrative privileges. Preferably, it gaphor's console should be possible to call from a fully packaged version so you don't need to build gtk to interact with it. This would make it a lot easier to develop scripts which interact with gaphor's data model.

### Describe the solution you'd like

Something along the lines of:
- Provide a .exe for the gaphor python console
- Provide some documentation of working with it (e.g., using it as a kernel in jupyter, setting it as a sphinx plugin)

### Describe alternatives you've considered

- Package a standalone version of command-line gaphor packages for interacting with/rendering the data model without having to install the gtk/UI pieces (if possible).

**Comment by danyeaw** [comment]
Hi @hulsed, to summarize this feature request, you are looking for these types of scripting capabilities like described here: https://docs.gaphor.org/en/latest/scripting.html, except you want to use it from a fully packaged version of Gaphor, is that right?

**Comment by hulsed** [comment]
I am referencing the scripting capabilities, yes, but I'm not necessarily tied to "calling fully packaged version" as a solution.

I think the key functionality would be being able to interact with gaphor command-line services without needing to build gtk. 

There are different ways of accomplishing that:

1. If I could call a packaged version of gaphor (and call its console), ostensibly there would be a decent glide path to interacting with the underlying data model from an external shell (python, etc). I put this up as there but I'm not sure it's really the most elegant approach and could mean anyone who is scripting has to write needless glue code. But it may (?) be easier to implement in terms of not changing much in the back-end.


2. Right now, if I clone gaphor and run `from gaphor.core.modeling import ElementFactory` from a script in the repository, in loading all the modules it will say I need `gaphas` (because it is being imported in `gaphor.core.modeling.diagram`). This puts me in a pickle--if I run a `pip install gaphas`, I then need to build gtk, even though it's not really necessary for the functionality I want to interact with. So my problem would be solved if elements of the library would be refactored so gui libraries are only loaded when needed. Potentially, this could be a more elegant solution, because then the user would be working in a single consistent python environment, but I'm not sure what would need to actually be done--it could be as little as changing some imports or as much as creating a separate repo/package for the data model.

**Comment by amolenaar** [comment]
@hulsed If you install Gaphor as a dependency in your own project, you should not need GTK per se. The scripting and Jupyter examples in the docs are executed that way.

You do need PyGObject, Pango (for text rendering), Cairo and Pycairo.

**Comment by hulsed** [comment]
Yes, I was grouping PyGObject and gtk there, since both are installed using gvsbuild on windows. It would be nice not to have to go through that process just to call a library that (I imagine?) doesn't actually need it for my use-case.

**Comment by hulsed** [comment]
Okay so there is a secret third way to solve my problem and that is installing dependencies via anaconda, which I have now figured out how to do.

I'd recommend adding some instructions for installing dependencies with anaconda in windows:

-----------------
Creating a development environment for gaphor in anaconda.

(1) Create new anaconda profile e.g., 'gaphor'

(2) Update anaconda in the environment

(3) Install packages

`conda install -c conda-forge gobject-introspection gtk4 pygobject pycairo hicolor-icon-theme adwaita-icon-theme`

Note: gtksourceview5, libadwaita are not available as anaconda packages. So you may (?) not be able to build/run the program in this environment.

(4) Set up your IDE(s)

`conda install ipython ipykernel`

(a) VSCode should work out-of-the box if it is already installed. Just set 'gaphor' as the kernel in your VSCode Profile or notebook

(b) Spyder may not be able to install under this environment (conda install spyder will reveal conflicts). 

Instead, you can do the following

`conda install spyder-kernels=2.4`

Then open spyder and set 'gaphor' as your python interpreter

**Comment by amolenaar** [comment]
> Yes, I was grouping PyGObject and gtk there, since both are installed using gvsbuild on windows. It would be nice not to have to go through that process just to call a library that (I imagine?) doesn't actually need it for my use-case.

Indeed it would certainly be nice to have PyGObject as an optional dependency altogether. I was not able to achieve this when I wanted to make Gaphor (as a library) work on readthedocs without GTK.

We do still need Cairo, though. Not sure if you can install that on windows without building it.

**Comment by amolenaar** [comment]
Th Conda way is interesting as well. With a tool like [Pixi](https://pixi.sh) we should be able to provide everything needed to run Gaphor with Jupyter.

**Comment by amolenaar** [comment]
BTW. Gaphor also supports running scripts from the CLI:

```
$ gaphor exec --help
usage: gaphor exec [-v | -q] [--profiler] [-h] script

Execute a python script from within Gaphor.

positional arguments:
  script         execute a script file and exit

options:
  -v, --verbose  enable verbose logging
  -q, --quiet    only show warning and error logging
  --profiler     run in profiler (cProfile)
  -h, --help     show this help message and exit
```

**Comment by hulsed** [comment]
@danyeaw I could do that, but not sure about time-frame since I would want to refine/double-check the approach (and I have a few things on my plate right now). Feel free to assign this to me in the mean-time so I can keep track of it.
